Blog: Announcing Apache Pulsar 5.0.0-M1 (preview release) (#1178)
apache/pulsar-site: blog/2026-06-23-announcing-apache-pulsar-5-0-m1.md

---
title: "Apache Pulsar 5.0.0-M1: A Preview of the Next Major Release"
author: Matteo Merli, Lari Hotari
date: 2026-06-23
---

> The Apache Pulsar community is voting on **Apache Pulsar 5.0.0-M1**, the first milestone on the road to Pulsar 5.0. It is a **preview**: an early build that puts the major new features of 5.0 in your hands so you can try them against real workloads and send feedback well ahead of the general-availability (GA) release. It is **not meant for production**.

Reviewer: the references to voting can be replaced about the release being already completed when the blog post gets published.

5.0 is a major release, and two changes stand out: **Scalable Topics** — a new kind of topic that grows and shrinks on its own — and the promotion of **Oxia** to Pulsar's recommended metadata store. This post walks through what's in M1 and how to try it.

<!--truncate-->

## A preview, built for feedback

> M1 is a milestone build, not a final release. We're publishing it early, during the community vote, for one reason: the changes in 5.0 are far-reaching, and we want real-world feedback before they're finalized for GA.

Reviewer: removing reference to voting phase
```suggestion
M1 is a milestone build, not a final release. We're publishing it early, for one reason: the changes in 5.0 are far-reaching, and we want real-world feedback before they're finalized for GA.
```

> So please run it on **non-production** clusters, exercise the new APIs, and tell us what works and what doesn't — open issues on [GitHub](https://github.com/apache/pulsar/issues) or start a thread on the [dev@pulsar.apache.org](https://pulsar.apache.org/contact/) mailing list. The feedback you give now directly shapes what 5.0 becomes at GA.

Reviewer: It would be useful to add a link to the docker compose guide for running 5.0.0-M1 together with Oxia.

Scalable topics are delivered by a family of proposals in 5.0:

- **[PIP-460](https://github.com/apache/pulsar/blob/master/pip/pip-460.md): Scalable Topics (Topics v5)** — the overall model: the segment DAG, range-based key routing, and design principles.
- **[PIP-468](https://github.com/apache/pulsar/blob/master/pip/pip-468.md): Scalable Topic Controller** — the broker-side controller that runs split/merge, assigns consumers, and pushes the live topology to clients.
- **[PIP-483](https://github.com/apache/pulsar/blob/master/pip/pip-483.md): Auto Split/Merge** — splits hot segments and merges cold ones automatically, tunable per broker, namespace, and topic.
- **[PIP-466](https://github.com/apache/pulsar/blob/master/pip/pip-466.md): New Java Client API (V5)** — a modern, sync-first Java client built for scalable topics.
- **[PIP-473](https://github.com/apache/pulsar/blob/master/pip/pip-473.md): Metadata-Driven Transactions** — transactions that survive segment splits and merges.
- **[PIP-475](https://github.com/apache/pulsar/blob/master/pip/pip-475.md): Regular-to-Scalable Migration** — convert an existing topic to a scalable topic in place, with no data copy.

> ### A new client API

Reviewer: Could we add a sentence somewhere in this section that current/previous client API remains supported, but scalable topics will only be available with the v5 API?

5.0 also **removes the etcd metadata store backend** ([PIP-462](https://github.com/apache/pulsar/blob/master/pip/pip-462.md)). ZooKeeper and Oxia are the supported backends going forward, so etcd users should migrate before upgrading.

## Other notable changes

- **IO connectors moved to a dedicated repository** ([PIP-465](https://github.com/apache/pulsar/blob/master/pip/pip-465.md)) — the built-in connectors now live and release independently of Pulsar core.
- **`javax.*` → `jakarta.*`** ([PIP-472](https://github.com/apache/pulsar/blob/master/pip/pip-472.md)) — Pulsar adopts the Jakarta EE namespace, a breaking change for code that touches the affected APIs.
- **Gradle build** ([PIP-463](https://github.com/apache/pulsar/blob/master/pip/pip-463.md)) — the build moves from Maven to Gradle, speeding up the development cycle for contributors.
- **Structured logging** ([PIP-467](https://github.com/apache/pulsar/blob/master/pip/pip-467.md)) — Pulsar adopts structured (slog-style) logging for richer, machine-parseable logs.

> - **Flexible networking** — multiple advertised addresses and smarter listener selection ([PIP-61](https://github.com/apache/pulsar/blob/master/pip/pip-61.md), [PIP-95](https://github.com/apache/pulsar/blob/master/pip/pip-95.md)) for multi-network deployments.

Reviewer: ```suggestion
- **Flexible networking** — improvement of multiple advertised addresses and smarter listener selection ([PIP-61](https://github.com/apache/pulsar/blob/master/pip/pip-61.md) (since 2.6), [PIP-95](https://github.com/apache/pulsar/blob/master/pip/pip-95.md)) (since 2.9) to support multi-network deployments in Pulsar 5.0.
```

## Try it and tell us what you think

> Once the vote passes and the artifacts are published, download 5.0.0-M1 and put it on a test cluster. Because this is a preview, we're especially keen to hear about:

Reviewer: ```suggestion
Deploy Pulsar 5.0 to a test cluster or test it locally with Docker Compose or Pulsar Standalone. Because this is a preview, we're especially keen to hear about:
```
